File: go/day10/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	input, _ := os.ReadFile("go/day10/input.txt")
	cpu := parse(string(input))
	sum := cpu.cycleSum([]int{20, 60, 100, 140, 180, 220})
	fmt.Printf("Part 1. %d\n", sum)
	cpu.cycleUntilCompletion()
	fmt.Printf("Part 2.\n%s\n", cpu.String())
}

func parse(input string) cpu {
	var instructions []instruction
	for _, line := range strings.Split(input, "\r\n") {
		words := strings.SplitN(line, " ", 2)
		instructionType := parseInstructionType(words[0])
		value := 0
		if len(words) > 1 {
			value, _ = strconv.Atoi(words[1])
		}
		instruction := instruction{
			instructionType: instructionType,
			value:           value,
			cyclesRemaining: instructionType.cyclesRequired(),
		}
		instructions = append(instructions, instruction)
	}
	return cpu{
		x:                  1,
		currentCycle:       0,
		currentInstruction: nil,
		instructions:       instructions,
		pc:                 0,
		crt:                []bool{},
	}
}

type cpu struct {
	x                  int
	currentCycle       int
	currentInstruction *instruction
	instructions       []instruction
	pc                 int
	crt                []bool
}

func (c *cpu) cycleSum(cycles []int) int {
	sum := 0
	for _, cycle := range cycles {
		sum += c.cycleTo(cycle)
	}
	return sum
}

func (c *cpu) cycleTo(n int) int {
	var x int
	for c.currentCycle < n {
		x = c.cycle()
	}
	return x * c.currentCycle
}

func (c *cpu) cycleUntilCompletion() int {
	var x int
	for c.pc < len(c.instructions)-1 {
		x = c.cycle()
	}
	return x * c.currentCycle
}

func (c *cpu) cycle() int {
	if c.currentInstruction == nil {
		c.currentInstruction = &c.instructions[c.pc]
		c.pc++
	}

	c.crt = append(c.crt, c.currentCycle%40 >= c.x-1 && c.currentCycle%40 <= c.x+1)

	c.currentInstruction.cyclesRemaining--
	xBeforeExecute := c.x
	if c.currentInstruction.cyclesRemaining == 0 {
		c.executeInstruction(*c.currentInstruction)
		c.currentInstruction = nil
	}
	c.currentCycle++
	return xBeforeExecute
}

func (c *cpu) executeInstruction(instruction instruction) {
	switch instruction.instructionType {
	case noop:
		return
	case addx:
		c.x += instruction.value
	}
}

func (c *cpu) String() string {
	var b strings.Builder
	for i, crt := range c.crt {
		if i%40 == 0 {
			b.WriteRune('\n')
		}
		if crt {
			b.WriteRune('#')
		} else {
			b.WriteRune('.')
		}
	}
	return b.String()
}

type instruction struct {
	instructionType instructionType
	value           int
	cyclesRemaining int
}

type instructionType int

const (
	noop instructionType = iota
	addx
)

func parseInstructionType(s string) instructionType {
	switch s {
	case "noop":
		return noop
	case "addx":
		return addx
	default:
		panic(fmt.Sprintf("invalid instruction type %q", s))
	}
}

func (t instructionType) cyclesRequired() int {
	switch t {
	case noop:
		return 1
	case addx:
		return 2
	default:
		panic("invalid instruction type")
	}
}

File: src/day10/main.rs
use std::fs;

// I got too annoyed again and remade it in Go.

fn main() {
    let input = fs::read_to_string("src/day10/input.txt").unwrap();
}

struct CPU {
    x: i64,
    cycle: usize,
    current_instruction: Option<Instruction>,
    instructions: Vec<Instruction>,
    pc: usize,
}

struct Instruction {
    instruction_type: InstructionType,
    value: i8,
    cycles_remaining: u8,
}

impl Instruction {
    fn new(instruction_type: InstructionType, value: Option<i8>) -> Self {
        let instruction_value = match value {
            Some(v) => v,
            None => 0,
        };
        let cycles_remaining = instruction_type.cycles_required();
        Instruction {
            instruction_type,
            value: instruction_value,
            cycles_remaining,
        }
    }
}

enum InstructionType {
    Noop,
    Addx,
}

impl InstructionType {
    fn new(s: &str) -> Self {
        match s {
            "noop" => InstructionType::Noop,
            "addx" => InstructionType::Addx,
            _ => panic!(),
        }
    }

    fn cycles_required(&self) -> u8 {
        match self {
            InstructionType::Noop => 1,
            InstructionType::Addx => 2,
        }
    }
}

impl CPU {
    fn new(input: &str) -> Self {
        let instructions = input
            .lines()
            .map(|l| {
                let words: Vec<&str> = l.split(' ').collect();
                let value = if words.len() > 1 {
                    Some(words[1].parse().unwrap())
                } else {
                    None
                };
                Instruction::new(InstructionType::new(words[0]), value)
            })
            .collect();
        CPU {
            x: 1,
            cycle: 0,
            current_instruction: None,
            instructions,
            pc: 0,
        }
    }

    fn cycle(&mut self, input: &str) {
        let cpu = self;
        if let Some(instruction) = cpu.current_instruction {
            cpu.current_instruction.unwrap().cycles_remaining -= 1;
            if instruction.cycles_remaining == 0 {
                cpu.execute_instruction(&instruction);
            }
        }
    }

    fn execute_instruction(&mut self, instruction: &Instruction) {
        match instruction.instruction_type {
            InstructionType::Addx => {
                self.x += instruction.value as i64;
            }
            InstructionType::Noop => {}
        }
    }
}
